clouddriver-kubernetes/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/caching/KubernetesV2ProviderSynchronizable.java

```diff
@@ -178,24 +176,35 @@ private void saveToCredentialsRepository(KubernetesNamedAccountCredentials accou
   }
 
   private void synchronizeKubernetesV2Provider(
-      Set<KubernetesNamedAccountCredentials<KubernetesV2Credentials>> allAccounts) {
+      Set<KubernetesNamedAccountCredentials<KubernetesV2Credentials>> newAndChangedAccounts) {
+
+    log.info(
+        "Synchronizing caching agents for {} new or changed V2 Kubernetes accounts.",
+        newAndChangedAccounts.size());
+
+    Set<String> accountNames = new HashSet<>();
 
     try {
-      for (KubernetesNamedAccountCredentials<KubernetesV2Credentials> credentials : allAccounts) {
+      for (KubernetesNamedAccountCredentials<KubernetesV2Credentials> credentials :
+          newAndChangedAccounts) {
+        accountNames.add(credentials.getName());
         List<Agent> newlyAddedAgents =
             kubernetesV2CachingAgentDispatcher.buildAllCachingAgents(credentials).stream()
                 .map(c -> (Agent) c)
                 .collect(Collectors.toList());
 
         log.info("Adding {} agents for account {}", newlyAddedAgents.size(), credentials.getName());
 
-        kubernetesV2Provider.addAllAgents(newlyAddedAgents);
+        kubernetesV2Provider.stageAllAgents(newlyAddedAgents);
       }
     } catch (Exception e) {
       log.warn("Error encountered scheduling new agents -- using old agent set instead", e);
-      kubernetesV2Provider.clearNewAgentSet();
+      kubernetesV2Provider.clearStagedAgents();
     }
 
+    // Remove existing agents belonging to changed accounts
+    ProviderUtils.unscheduleAndDeregisterAgents(accountNames, catsModule);
```

Comment:
If an exception occurs while building the agent (line 192), it will be caught and all agents will then be unregistered. I think we should 1) try/catch in the loop 2) not unregister/unschedule agents that had been previously scheduled unless they're explicitly deleted.

Reply:
Good catch